app/src/main/java/com/github/polybooks/core/database/DatabaseAPI.kt

```diff
@@ -0,0 +1,176 @@
+package com.github.polybooks.core.database
+
+import com.github.polybooks.core.*
+import java.util.concurrent.CompletableFuture
+
+/**
+ * Provides the API for a database. Allows to retrieve information in the database
+ * about Books, Sales.
+ * */
+interface Database {
+
+    /**
+     * Creates a new query for Books. It originally matches all books.
+     * */
+    fun queryBooks() : BookQuery
+
+    /**
+     * Create a new query for Sales. It originally matches all sales.
+     * */
+    fun querySales() : SaleQuery
+
+    //TODO we could or could not include those functions (and more) in the interface. Should we?
+
+    fun listAllBooks() : CompletableFuture<List<Book>> = TODO("It can be implemented from the previous functions")
+
+    fun listAllSales() : CompletableFuture<List<Sale>> = TODO("It can be implemented from the previous functions")
+
+    /**
+     * Get data about a Book from the database given it's ISBN13
+     * */ //TODO define how to handle an invalid isbn13
+    fun getBook(isbn13 : Long) : CompletableFuture<Book> = TODO("It can be implemented from the previous functions")
+
+
+    /**
+     * A method for getting books by batches of at most N books. The batches are indexed by ordered pages.
+     * @param numberOfBooks The maximum number of books per page
+     * @param page The index of the page
+     * @param ordering The ordering for the pages and books within the pages (see {@link BookOrdering})
+     * */
+    fun getNBooks(numberOfBooks : Int, page : Int, ordering : BookOrdering) : CompletableFuture<List<Book>>
+        = TODO("It can be implemented from the previous functions")
+
+    //TODO need to add methods to modify the database, create entries.
+
+}
+
+/**
+ * A BookQuery is a builder for a query to the database that will yield Books.
+ * Most methods return themselves for function chaining
+ * */
+interface BookQuery {
+
+    /**
+     * Execute the query and return all the results in a Future.
+     * */
+    fun getAll() : CompletableFuture<List<Book>>
+
+    /**
+     * Get the results in batches of at most n books.
+     * */
+    fun getN(n : Int, page : Int) : CompletableFuture<List<Book>>
+
+    /**
+     * Get how many Books match this query
+     * */
+    fun getCount() : CompletableFuture<Int>
+
+    /**
+     * Set this query to not include books that satisfy the given interests.
```

Comment:
I can't think of a practical use case where we want to get from the db the books that the user **don't ** care about ?

Reply:
hmmm, I thought it would create confusion. I need to find another wording.
A query would initially match all interests. That function would remove the specified interests from the list of interests that are matched. Not exclude them, but remove them from the list of interests to include.

I need to think more about this. Do we really need two functions to add and remove interests, should it be default match all interests or none?